query/queryable.d.ts

```diff
@@ -0,0 +1,146 @@
+ */
+export type Algebra = any;
+
+/**
+ * Generic query engine interfaces.
+ * It allow engines to return any type of result object for any type of query.
+ * @param QueryFormatTypesAvailable The format of the query, either string or algebra object.
+ * @param SourceType The allowed sources over which queries can be executed.
+ * @param SupportedMetadataType The allowed metadata types.
+ * @param QueryType The allowed query types.
+ * @param QueryStringContextType Type of the string-based query context.
+ * @param QueryAlgebraContextType Type of the algebra-based query context.
+ */
+export interface Queryable<
+  QueryFormatTypesAvailable extends string | Algebra,
+  SourceType,
+  SupportedMetadataType,
+  QueryType extends Query<SupportedMetadataType>,
+  QueryStringContextType extends QueryStringContext<SourceType>,
+  QueryAlgebraContextType extends QueryAlgebraContext<SourceType>,
+> {
+  /**
+   * Initiate a given query.
+   *
+   * This will produce a future to a query result, which has to be executed to obtain the query results.
+   *
+   * This can reject given an unsupported or invalid query.
+   *
+   * @see Query
+   */
+  query<QueryFormatType extends QueryFormatTypesAvailable>(
+    query: QueryFormatType,
+    context?: QueryFormatType extends string ? QueryStringContextType : QueryAlgebraContextType,
+  ): Promise<QueryType>;
+}
+
+/**
+ * SPARQL-constrainted query interface.
+ *
+ * This interface guarantees that result objects are of the expected type as defined by the SPARQL spec.
+ */
+export type SparqlQueryable<
+  QueryFormatTypesAvailable extends string | Algebra,
+  SourceType,
+  QueryStringContextType extends QueryStringContext<SourceType>,
+  QueryAlgebraContextType extends QueryAlgebraContext<SourceType>,
+  SupportedResultType,
+> = unknown
+  & (SupportedResultType extends BindingsResultSupport ? {
+  queryBindings<QueryFormatType extends QueryFormatTypesAvailable>(
+    query: QueryFormatType,
+    context?: QueryFormatType extends string ? QueryStringContextType : QueryAlgebraContextType,
+  ): Promise<ResultStream<Bindings>>;
+} : unknown)
+  & (SupportedResultType extends BooleanResultSupport ? {
+  queryBoolean<QueryFormatType extends QueryFormatTypesAvailable>(
+    query: QueryFormatType,
+    context?: QueryFormatType extends string ? QueryStringContextType : QueryAlgebraContextType,
+  ): Promise<boolean>;
+} : unknown)
+  & (SupportedResultType extends QuadsResultSupport ? {
+  queryQuads<QueryFormatType extends QueryFormatTypesAvailable>(
+    query: QueryFormatType,
+    context?: QueryFormatType extends string ? QueryStringContextType : QueryAlgebraContextType,
+  ): Promise<ResultStream<RDF.Quad>>;
+} : unknown)
+  & (SupportedResultType extends VoidResultSupport ? {
+  queryVoid<QueryFormatType extends QueryFormatTypesAvailable>(
+    query: QueryFormatType,
+    context?: QueryFormatType extends string ? QueryStringContextType : QueryAlgebraContextType,
+  ): Promise<void>;
+} : unknown)
```

Comment:
There was also a long discussion about this one here: https://github.com/rdfjs/query-spec/pull/7

Eventually, we reached a consensus about the current methods.
(I don't remember the exact argumentation, perhaps @jacoscaz or @gsvarovsky knows?)

Reply:
I'm definitely in favour of more fluent method names. My only reservation is that these methods obtain a query object/proxy but don't execute it. But at this point that's probably splitting hairs.

Reply:
What does it mean that they return a proxy?

Reply:
If we all agree on having these methods return results directly, without an intermediate `Query` object, then we might also agree on changing their names as per @tpluscode's suggestion. Anyone against this?

Reply:
I just remembered a (significant) reason to keep the `query...` methods: interface discovery through autocompletion.

Since the `Queryable` interface already exposes a `query` method, developers that don't need the API docs might start typing `query...` in their editor. Smart editors would then immediately also suggest `queryBindings`, `queryQuads`, ... This would help developers get directed to the simpler query methods from the `SparqlQueryable` interface _if_ they are available.

If we would go for `select`, `construct`, etc, then we have to take into account the fact that `construct` might cause some confusion, since we don't have a 1-1 mapping with the SPARQL query clauses. Namely, `construct` would be used for both `CONSTRUCT` and `DESCRIBE` query clauses, which may not be obvious for newcomers.

Reply:
Well, if were to go with renaming to `select`, `construct`, ...  then we should also have a `describe`, which in the majority of cases would probably call `construct` internally. I guess on one side we would favor discovery by autocompletion, on the other side we would favor discovery by intuition. I can see the point in both approaches, happy to go with either.

Reply:
Another (minor) argument in favor of `query...` is the alignment with the `resultType` field of `Query`.

Reply:
All told, there seem to be more arguments in favor of keeping these methods as they are. I think we can resolve this and proceed with merging.